File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/players/shared/SharedPlayerMerger.kt
package com.blaze.flutterblazesdk.players.shared

import android.content.Context
import com.blaze.blazesdk.style.players.BlazeFirstTimeSlideCTAStyle
import com.blaze.blazesdk.style.players.BlazeFirstTimeSlideInstructionStyle
import com.blaze.blazesdk.style.players.BlazeFirstTimeSlideTextStyle
import com.blaze.blazesdk.style.players.BlazePlayerButtonCustomImageStates
import com.blaze.blazesdk.style.players.BlazeSeekBarStyle
import com.blaze.blazesdk.style.shared.models.blazeDp
import com.blaze.flutterblazesdk.utils.parsing.safeParseColor
import com.blaze.flutterblazesdk.utils.parsing.toColorResId
import com.blaze.flutterblazesdk.utils.parsing.toFontResId
import com.blaze.flutterblazesdk.utils.parsing.toImageResId

fun BlazeFirstTimeSlideInstructionStyle.mergedWith(
        customization: BlazeReactFirstTimeSlideInstructionStyle?,
        context: Context
): BlazeFirstTimeSlideInstructionStyle {
    customization ?: return this

    val merged = this
    merged.headerText = this.headerText.mergedWith(customization.headerText, context)
    merged.descriptionText = this.descriptionText.mergedWith(customization.descriptionText, context)
    return merged
}

fun BlazeFirstTimeSlideTextStyle.mergedWith(
        customization: BlazeReactFirstTimeSlideTextStyle?,
        context: Context
): BlazeFirstTimeSlideTextStyle {
    customization ?: return this

    val merged = this
    merged.text = customization.text ?: this.text
    customization.textColor?.colorFileName?.toColorResId(context)?.let {
        merged.textColorResId = it
    }
    merged.fontResId = customization.font?.toFontResId(context)
    merged.textSize = customization.textSize ?: this.textSize

    return merged
}

fun BlazeFirstTimeSlideCTAStyle.mergedWith(
        customization: BlazeReactFirstTimeSlideCTAStyle?,
        context: Context
): BlazeFirstTimeSlideCTAStyle {
    customization ?: return this

    val merged = this
    merged.title = customization.title ?: this.title
    merged.cornerRadius = customization.cornerRadius?.blazeDp ?: this.cornerRadius
    customization.backgroundColor?.colorFileName?.toColorResId(context)?.let {
        merged.backgroundColor = it
    }
    customization.textColor?.colorFileName?.toColorResId(context)?.let {
        merged.textColorResId = it
    }

    merged.fontResId = customization.font?.toFontResId(context)
    merged.textSize = customization.textSize ?: this.textSize
    return merged
}

fun BlazeSeekBarStyle.mergedWith(
        customization: BlazeReactSeekBarStyle?,
        context: Context
): BlazeSeekBarStyle {
    customization ?: return this

    val merged = this

    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.backgroundColor = safeParseColor(customization.backgroundColor) ?: this.backgroundColor
    merged.progressColor = safeParseColor(customization.progressColor) ?: this.progressColor
    merged.height = customization.height?.blazeDp ?: this.height
    merged.cornerRadius = customization.cornerRadius?.toInt()?.blazeDp ?: this.cornerRadius
    merged.thumbColor = safeParseColor(customization.thumbColor) ?: this.thumbColor
    merged.thumbImageResId = customization.thumbImage?.toImageResId(context)
    merged.thumbSize = customization.thumbSize?.blazeDp ?: this.thumbSize
    merged.isThumbVisible = customization.isThumbVisible ?: this.isThumbVisible

    return merged
}

fun <T : com.blaze.blazesdk.style.players.IPlayerItemButtonStyle> T.mergeButtonThemes(
    customization: BlazeReactPlayerButtonStyle?,
    context: Context
): T {
    customization ?: return this

    customization.width?.let { this.width = it.blazeDp }
    customization.height?.let { this.height = it.blazeDp }
    safeParseColor(customization.color)?.let { this.color = it }
    customization.isVisible?.let { this.isVisible = it }
    customization.scaleType?.let {
        this.scaleType = android.widget.ImageView.ScaleType.valueOf(it)
    }
    customization.isVisibleForAds?.let { this.isVisibleForAds = it }
    customization.customImage?.let {
        this.customImage = this.customImage.mergedWith(it, context)
    }

    return this
}

fun BlazePlayerButtonCustomImageStates?.mergedWith(
    customization: BlazeReactPlayerButtonCustomImageStates,
    context: Context
): BlazePlayerButtonCustomImageStates? {
    val imageUnselectedPath =
        customization.unselectedImage?.toImageResId(context) ?: return this
    return BlazePlayerButtonCustomImageStates(
        imageUnselectedPathResId = imageUnselectedPath,
        imageSelectedPathResId = customization.selectedImage?.toImageResId(context)
    )
}

File: blaze_flutter_sdk/ios/Sources/Players/Shared/SharedPlayerMerger.swift
//
//  SharedPlayerMerger.swift
//  blaze_flutter_sdk
//
//

import BlazeSDK
import Foundation

extension BlazeFirstTimeSlideInstructionStyle {
    func mergedWith(_ customization: BlazeReactFirstTimeSlideInstructionStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.descriptionText = merged.descriptionText.mergedWith(customization.descriptionText)
        merged.headerText = merged.headerText.mergedWith(customization.headerText)
        return merged
    }
}

extension BlazeFirstTimeSlideTextStyle {
    func mergedWith(_ customization: BlazeReactFirstTimeSlideTextStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.text = customization.text ?? merged.text
        merged.textColor =
            UIColor(hexString: customization.textColor?.colorName) ?? merged.textColor
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        return merged
    }
}

extension BlazeFirstTimeSlideCTAStyle {
    func mergedWith(_ customization: BlazeReactFirstTimeSlideCTAStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor?.colorName) ?? merged.backgroundColor
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.title = customization.title ?? merged.title
        merged.textColor =
            UIColor(hexString: customization.textColor?.colorName) ?? merged.textColor
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        return merged
    }
}

extension BlazeSeekBarStyle {
    func mergedWith(_ customization: BlazeReactSeekBarStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.progressColor =
            UIColor(hexString: customization.progressColor) ?? merged.progressColor
        merged.height = customization.height ?? merged.height
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.thumbColor = UIColor(hexString: customization.thumbColor) ?? merged.thumbColor
        merged.thumbImage = UIImage.imageWith(customization.thumbImage) ?? merged.thumbImage
        merged.thumbSize = customization.thumbSize ?? merged.thumbSize
        merged.isThumbVisible = customization.isThumbVisible ?? merged.isThumbVisible
        return merged
    }
}
